### ISSUE TITLE
Tiling on very large/ultra wide monitors

### ISSUE THREAD

**Issue by Theaninova** [pop-os/cosmic-epoch#410]
I'm blown away by the pre-alpha so far, however the one thing that I'm really missing is how tiling behaves on very large (>40") or ultra wide monitors (I'll refer to these just as "very large" for simplicity).

## The issue with normal (auto-)tiling on very large screens

<details><summary>Text tends to remain legible up to around 16-20cm or so in width at standard sizes, and apps are designed around that.</summary>

For example
* Code formatters will usually cap line lengths at 80-120 characters max, leaving the rest of the space blank
* Text on websites rarely goes beyond that width
* Non-limited apps tend to be even worse, as it is close to impossible to keep track to where you were looking when a line is a meter long

It's not just an app issue, there's only so much content you can put in side bars.

</details>

Traditional (auto-)tiling places content in two columns, in case there's only a single window that one will span the entire monitor.

This however creates a significant issue with large monitors.

* <details><summary>A single window spans the entire monitor</summary> This not just leaves most of the content blank, it also puts focus on the left most part of the screen, making you turn your head uncomfortably to that direction. Even apps and sites you'd usually associate with being optimized for large monitors tend to fall apart, web browsers, IDEs, etc.</details>
* <details><summary>A two column layout has no central focus for your eyes</summary> you need to turn your head left or right to actually see the content and interact with it. This is fine on small monitors, but on very large ones it's a significant ergonomic problem to have your main content off to the sides. Imagine having a dual monitor setup, with the seam right in the middle.</details>

In essence, it requires you to always have your head turned to one direction.

## What I did so far in other tiling WMs

The solution for me so far was the "master stack" layout with three columns and keeping gaps if there's only one window. Main content is kept at optimal reading width, while also remaining centered and comfortable to look at.

I'd probably be really happy with that as well, but why settle for something good when you can have something better :)

<details>
<summary>Hyprland config</summary>

```
general {
  layout = master
}

master {
  orientation = center
  always_center_master = true
  mfact = 0.4
}
```

</details>

<details>
<summary>For reference this is how my Hyprland layout looks like</summary>
<img src="https://github.com/pop-os/cosmic-epoch/assets/19289296/ef7c5b2f-19e8-4d5b-a5ae-b16b62e1af41">
</details>

## My dream solution
This would probably be "virtual monitors"
* <details><summary>Virtual monitors would be edited with something akin to KDE's tiling editor. The bar, wallpaper and dock would span all virtual monitors.</summary> In case you don't know what KDE's tiling editor looks like: <img src="https://github.com/pop-os/cosmic-epoch/assets/19289296/d2cc9e25-17c2-403c-b276-73f6de5254d6"></details>
* Each virtual monitor can then switch workspaces independently from the others, just like a physical monitor (or as you set in the settings)
* Fullscreen would span all virtual monitors, but still belong to one of the virtual monitors. Switching workspaces would switch workspaces on the virtual monitor and subsequently hide the fullscreen window.
### Bonus
* <details><summary>A workspace would ideally have the option to push away new windows to different monitors if there's already an active one</summary>Maybe as part of a window being full-screen or maximized? This would be beneficial for keeping your central virtual monitor clean and maybe even to avoid hiding windows behind fullscreen/maximized windows?</details>
* <details><summary>Each workspace on a virtual monitor would have the option to be either tiling or floating, just like it is now.</summary>This could be used for example to dedicate a small area somewhere as a scratchpad with floating windows.</details>

**Comment by autom4il** [comment]
I have achieved almost the same in my WM but rather than having three columns I prefer working with two. Anyhow, I just want to say that I'm interested in this feature too.

**Comment by dkirchhof** [comment]
Will it be implemented at some point?
A master-stack-layout with configurable column widths (like in hyprland) would be amazing.

**Comment by mike-lloyd03** [comment]
I had a similar request in pop-shell to enable a maximum window width. I opened a PR for it which got merged. (https://github.com/pop-os/shell/pull/1762)

This prevented windows from occupying the full width of the display and the user could specify a maximum width in the extension settings. I could look into implementing something similar here if there's interest.

**Comment by Frosthaven** [comment]
Just adding myself as another interested party. I'm looking for either the ability to center window containers to a configured maximum width when not full-screen, or the ability to have monitor-specific workspace x-padding when there's a single window.

Currently the only thing to keep me from staring off to the left after work is opening programs I don't need on either side of the one I do want.

**Comment by skewballfox** [comment]
Could this be achieved with smarter tiling rather than making emulating multiple monitors virtually? Like say you have an ultrawide monitor. Launch one app, it takes the full screen, 2-3 the screen just divides normally. past 3 it starts treating those regions as independent, and spawing them horizontally.

**Comment by clintre** [comment]
This is honestly a must for me. I currently use KDE + Kzones as I need 3 columns for my preferred workflow. I use multiple monitors, with my main one being a 49" inch 32:9 monitor. Would love to see this on Cosmic.

**Comment by etrigan63** [comment]
Ultrawide monitor support is absolutely necessary when tiling.

**Comment by bhh32** [comment]
I use an ultra wide as well and have no issues with tiling on it.

**Comment by clintre** [comment]
> I use an ultra wide as well and have no issues with tiling on it.

Are you able to have 3 columns?

**Comment by bhh32** [comment]
Yes, if I arrange them that way.

**Comment by mikemayuare** [comment]
Hyprland has a setting to set the default aspect ratio for single windows, that way a single window is nice and centered and as soon a seciond window is opened, everything tiles as expected.

**Comment by Frosthaven** [comment]
> Hyprland has a setting to set the default aspect ratio for single windows, that way a single window is nice and centered and as soon a seciond window is opened, everything tiles as expected.

This is my hope for any implementation if it is taken on. Super-ultrawide tiling is perfect for work, but not so great for browsing on a single window since websites cap out their widths. I've gotten in the habit of keeping a terminal open with my browser and looking left all the time, which isn't ideal.

**Comment by mikemayuare** [comment]
> This is my hope for any implementation if it is taken on. Super-ultrawide tiling is perfect for work, but not so great for browsing on a single window since websites cap out their widths. I've gotten in the habit of keeping a terminal open with my browser and looking left all the time, which isn't ideal.

Agree, also for coding as many text editor put the text zone to the left.

**Comment by HenrikDK2** [comment]
Hyprland does have master layout, I'm used to having a 16:9 ratio window in the middle of the screen (1440p), and two smaller windows to the side.

Great for productivity, and gaming when using discord and music player to the side.

Having one big browser window is not ideal for 49" monitor, the same goes with some games, I constantly have to fiddle between floating and tilling mode in Cosmic to get a decent setup in one workspace. If it could be possible to have some way of redefining the cosmic tilling default layout that would be ideal.

**Comment by mike-lloyd03** [comment]
I've started work on this. It currently limits the maximum width of each window to a specified value so no single window will span the entire screen. The value is currently hardcoded to 2400 pixels. I still haven't figured out how to connect the configuration option to the actual maximum width value in `src/shell/layout/tiling/mod.rs`. But you can check it out here if you're interested in contributing.

https://github.com/mike-lloyd03/cosmic-comp

Update: PR opened
